File: firmware/hardware_core/src/framing.cpp
#include "framing.h"

namespace orion {

namespace {
constexpr uint16_t POLINOMIO = 0x1021;
constexpr uint16_t VALOR_INICIAL = 0xFFFF;
}  // namespace

uint16_t crc16(const uint8_t* dados, size_t tamanho) {
  uint16_t crc = VALOR_INICIAL;
  for (size_t i = 0; i < tamanho; i++) {
    crc ^= static_cast<uint16_t>(dados[i]) << 8;
    for (uint8_t bit = 0; bit < 8; bit++) {
      if (crc & 0x8000) {
        crc = (crc << 1) ^ POLINOMIO;
      } else {
        crc = crc << 1;
      }
    }
  }
  return crc;
}

void enviarQuadro(Stream& porta, const uint8_t* payload, size_t tamanho) {
  uint16_t crc = crc16(payload, tamanho);
  uint8_t crcBytes[2] = {static_cast<uint8_t>(crc >> 8), static_cast<uint8_t>(crc & 0xFF)};

  porta.write(FRAME_STX);
  for (size_t i = 0; i < tamanho; i++) {
    uint8_t b = payload[i];
    if (b == FRAME_STX || b == FRAME_ETX || b == FRAME_ESC) {
      porta.write(FRAME_ESC);
      porta.write(static_cast<uint8_t>(b ^ FRAME_XOR_ESCAPE));
    } else {
      porta.write(b);
    }
  }
  for (uint8_t b : crcBytes) {
    if (b == FRAME_STX || b == FRAME_ETX || b == FRAME_ESC) {
      porta.write(FRAME_ESC);
      porta.write(static_cast<uint8_t>(b ^ FRAME_XOR_ESCAPE));
    } else {
      porta.write(b);
    }
  }
  porta.write(FRAME_ETX);
}

void enviarQuadro(Stream& porta, const String& payload) {
  enviarQuadro(porta, reinterpret_cast<const uint8_t*>(payload.c_str()), payload.length());
}

bool DecodificadorQuadro::alimentar(uint8_t byte) {
  if (byte == FRAME_STX) {
    // Reinicia mesmo no meio de um quadro - permite ressincronizar apos
    // ruido/quadro incompleto no link serial.
    _emQuadro = true;
    _escapando = false;
    _tamanho = 0;
    return false;
  }

  if (!_emQuadro) {
    return false;
  }

  if (byte == FRAME_ETX && !_escapando) {
    _emQuadro = false;
    return finalizarQuadro();
  }

  if (byte == FRAME_ESC && !_escapando) {
    _escapando = true;
    return false;
  }

  if (_escapando) {
    byte ^= FRAME_XOR_ESCAPE;
    _escapando = false;
  }

  if (_tamanho < CAPACIDADE) {
    _buffer[_tamanho++] = byte;
  } else {
    // buffer estourado - descarta o quadro (Cap 14 s.5: erro isolado, nao crash)
    _emQuadro = false;
    _quadrosInvalidos++;
  }
  return false;
}

bool DecodificadorQuadro::finalizarQuadro() {
  if (_tamanho < 2) {
    _quadrosInvalidos++;
    return false;
  }
  size_t tamanhoPayload = _tamanho - 2;
  uint16_t crcRecebido = (static_cast<uint16_t>(_buffer[tamanhoPayload]) << 8) |
                         static_cast<uint16_t>(_buffer[tamanhoPayload + 1]);
  uint16_t crcEsperado = crc16(_buffer, tamanhoPayload);
  if (crcRecebido != crcEsperado) {
    _quadrosInvalidos++;
    return false;
  }
  _tamanho = tamanhoPayload;  // deixa so o payload visivel em buffer()/tamanho()
  return true;
}

}  // namespace orion

File: src/orion/communication/framing.py
"""Camada de enquadramento (Cap 14 secao 3).

Serial (Raspberry<->Arduino): delimitadores STX/ETX, escape de bytes
(byte-stuffing) e CRC16 - necessario porque a porta serial e um fluxo de
bytes sem nocao de mensagem. CRC invalido -> descarte silencioso (Cap 14
secao 5); quem chama decide se pede retransmissao (NACK).

TCP (Notebook<->Raspberry): o proprio Cap 14 diz que no TCP "o framing
delimita mensagens no stream" - sem escape/CRC, porque o TCP ja garante
entrega ordenada e integra dos bytes. Usamos prefixo de tamanho (4 bytes),
que delimita sem ambiguidade (evita basear o parsing em varredura de
delimitador, que ja causou um bug de enquadramento no projeto irmao
Sentinela X quando o escape nao existia).
"""
from __future__ import annotations

from orion.communication.crc import crc16

STX = 0x02
ETX = 0x03
ESC = 0x1B
_XOR_ESCAPE = 0x20

_BYTES_ESPECIAIS = (STX, ETX, ESC)

_PREFIXO_TAMANHO_TCP = 4


def codificar_serial(payload: bytes) -> bytes:
    """STX + payload_escapado + crc16_escapado + ETX."""
    crc = crc16(payload).to_bytes(2, "big")
    corpo = bytearray()
    for byte in payload + crc:
        if byte in _BYTES_ESPECIAIS:
            corpo.append(ESC)
            corpo.append(byte ^ _XOR_ESCAPE)
        else:
            corpo.append(byte)
    return bytes([STX]) + bytes(corpo) + bytes([ETX])


class DecodificadorSerial:
    """Decodificador stateful: alimente byte a byte conforme chegam da porta serial."""

    def __init__(self) -> None:
        self._em_quadro = False
        self._escapando = False
        self._buffer = bytearray()
        self.quadros_invalidos = 0

    def alimentar(self, byte: int) -> bytes | None:
        """Processa um byte. Retorna o payload decodificado quando um quadro
        valido se completa, ou None caso contrario (quadro ainda incompleto,
        ou byte de ruido fora de um quadro)."""
        if byte == STX:
            # Reinicia mesmo se ja estava no meio de um quadro - permite
            # ressincronizar apos ruido/quadro incompleto no link serial.
            self._em_quadro = True
            self._escapando = False
            self._buffer = bytearray()
            return None

        if not self._em_quadro:
            return None

        if byte == ETX and not self._escapando:
            self._em_quadro = False
            return self._finalizar_quadro()

        if byte == ESC and not self._escapando:
            self._escapando = True
            return None

        if self._escapando:
            byte ^= _XOR_ESCAPE
            self._escapando = False

        self._buffer.append(byte)
        return None

    def alimentar_bytes(self, dados: bytes) -> list[bytes]:
        """Conveniencia para alimentar varios bytes de uma vez (ex.: em testes)."""
        quadros = []
        for byte in dados:
            quadro = self.alimentar(byte)
            if quadro is not None:
                quadros.append(quadro)
        return quadros

    def _finalizar_quadro(self) -> bytes | None:
        if len(self._buffer) < 2:
            self.quadros_invalidos += 1
            return None
        payload, crc_recebido = bytes(self._buffer[:-2]), bytes(self._buffer[-2:])
        if crc_recebido != crc16(payload).to_bytes(2, "big"):
            self.quadros_invalidos += 1
            return None
        return payload


def codificar_tcp(payload: bytes) -> bytes:
    """Prefixo de tamanho (4 bytes, big-endian) + payload."""
    return len(payload).to_bytes(_PREFIXO_TAMANHO_TCP, "big") + payload


class DecodificadorTcp:
    """Decodificador stateful para o framing por prefixo de tamanho do TCP."""

    def __init__(self) -> None:
        self._buffer = bytearray()

    def alimentar(self, dados: bytes) -> list[bytes]:
        """Adiciona bytes recebidos do socket e retorna as mensagens completas."""
        self._buffer.extend(dados)
        mensagens = []
        while True:
            if len(self._buffer) < _PREFIXO_TAMANHO_TCP:
                break
            tamanho = int.from_bytes(self._buffer[:_PREFIXO_TAMANHO_TCP], "big")
            fim = _PREFIXO_TAMANHO_TCP + tamanho
            if len(self._buffer) < fim:
                break
            mensagens.append(bytes(self._buffer[_PREFIXO_TAMANHO_TCP:fim]))
            del self._buffer[:fim]
        return mensagens
